frontends/halo2/src/backend/llzk/factory.rs

```diff
@@ -1,96 +1,87 @@
-    let loc = Location::new(context, &filename, 0, 0);
-
-    r#struct::field(loc, name, FeltType::new(context), false, public)
-}
-
-fn struct_type<'c>(context: &'c Context, name: &str) -> Type<'c> {
-    StructType::from_str(context, name).into()
-}
-
-struct Field {
-    name: Cow<'static, str>,
-    public: bool,
-}
-
-impl From<(&'static str, bool)> for Field {
-    fn from(value: (&'static str, bool)) -> Self {
-        Self {
-            name: Cow::Borrowed(value.0),
-            public: value.1,
-        }
-    }
-}
-
-impl Field {
-    pub fn renamed(mut self, f: impl FnOnce(&str) -> String) -> Self {
-        let new = f(self.name.as_ref());
-        self.name = Cow::Owned(new);
-        self
-    }
-
-    pub fn name(&self) -> &str {
-        self.name.as_ref()
-    }
-
-    pub fn is_public(&self) -> bool {
-        self.public
-    }
-}
-
 pub struct StructIO {
     private_inputs: usize,
     public_inputs: usize,
     private_outputs: usize,
     public_outputs: usize,
 }
 
-macro_rules! field_iter {
-    ($name:ident, $base:expr) => {
-        fn $name(&self) -> impl Iterator<Item = Field> {
-            iter::repeat($base)
-                .map(Field::from)
-                .zip(0..self.$name)
-                .map(|(f, n)| f.renamed(|name| format!("{name}_{n}")))
-        }
-    };
-}
-
 impl StructIO {
-    fn fields(&self) -> impl IntoIterator<Item = Field> {
-        self.private_inputs()
-            .chain(self.private_outputs())
-            .chain(self.public_outputs())
+    fn fields<'c>(
+        &self,
+        context: &'c Context,
+        header: &str,
+    ) -> impl Iterator<Item = Result<FieldDefOp<'c>, LlzkError>> {
+        let public_filename = format!("struct {header} | public outputs");
+        let private_filename = format!("struct {header} | private outputs");
+        std::iter::repeat_n(true, self.public_outputs)
+            .enumerate()
+            .chain(std::iter::repeat_n(false, self.private_outputs).enumerate())
+            .map(move |(n, public)| {
+                let filename = if public {
+                    &public_filename
+                } else {
+                    &private_filename
+                };
+                (public, Location::new(context, filename, n, 0))
+            })
+            .enumerate()
+            .map(|(n, (public, loc))| {
+                let name = format!("out_{n}");
+                r#struct::field(loc, &name, FeltType::new(context), false, public)
+            })
     }
 
-    field_iter!(private_inputs, ("in", false));
-    field_iter!(private_outputs, ("out", false));
-    field_iter!(public_outputs, ("out", true));
+    pub fn args<'c>(
+        &self,
+        ctx: &'c Context,
+        is_main: bool,
+        struct_name: &str,
+    ) -> Vec<(Type<'c>, Location<'c>)> {
+        let public_filename = format!("struct {struct_name} | public inputs");
+        let private_filename = format!("struct {struct_name} | private inputs");
```

Comment:
Maybe we should extract the "filename" creation out into a separate function in case we decide to change the format on this.